testutils/gke/client.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gke
+
+import (
+	"fmt"
+
+	container "google.golang.org/api/container/v1beta1"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+)
+
+// SDKOperations wraps GKE SDK related functions
+type SDKOperations interface {
+	CreateCluster(string, string, *container.CreateClusterRequest) error
+	CreateClusterAsync(string, string, *container.CreateClusterRequest) (*container.Operation, error)
+	DeleteCluster(string, string, string) error
+	DeleteClusterAsync(string, string, string) (*container.Operation, error)
+	GetCluster(string, string, string) (*container.Cluster, error)
+	GetOperation(string, string, string) (*container.Operation, error)
+}
+
+// sdkClient Implement SDKOperations
+type sdkClient struct {
+	*container.Service
+}
+
+// NewSDKClient returns an SDKClient that implements SDKOperations
+func NewSDKClient() (SDKOperations, error) {
+	ctx := context.Background()
+	c, err := google.DefaultClient(ctx, container.CloudPlatformScope)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Google client: '%v'", err)
+	}
+
+	containerService, err := container.New(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create container service: '%v'", err)
+	}
+	return &sdkClient{containerService}, nil
+}
+
+// CreateCluster creates a new GKE cluster, and wait until it finishes or timeout or there is an error.
+func (gsc *sdkClient) CreateCluster(
+	project, location string,
+	rb *container.CreateClusterRequest,
+) error {
+	op, err := gsc.CreateClusterAsync(project, location, rb)
+	if err == nil {
+		err = Wait(gsc, project, location, op.Name, creationTimeout)
+	}
+	return err
+}
+
+// CreateClusterAsync creates a new GKE cluster asynchronously.
+func (gsc *sdkClient) CreateClusterAsync(
```

Comment:
I have a thought here to make this method smarter: what if we save a `lastOpName` var, and have an API `WaitLastOperation` which waits for latest kicked off aync calls?